### DISCUSSION TITLE
Problem Building rtl_433 for Windows

### DISCUSSION THREAD

**Discussion by MaxFun** [merbanan/rtl_433#2328] [General]
Tried to building rtl_433 for Windows using the instruction on https://github.com/merbanan/rtl_433/blob/master/docs/BUILDING.md
First tried building using the instruction for MinGW-w64, but the download does not contain the file "FindLibUSB.cmake" which needed to be modified.

Next I followed the instruction for building rtl_433 with Visual Studio 2017. I installed PothosSDR (PothosSDR-2021.07.25-vc16-x64.exe) then cloned https://github.com/merbanan/rtl_433.git. VS output had no errors indicated

VS 2017 output
1> Command line: C:\PROGRAM FILES (X86)\MICROSOFT VISUAL STUDIO\2017\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\CMake\bin\cmake.exe  -G "Ninja" -DCMAKE_INSTALL_PREFIX:PATH="C:\Users\MySelf\CMakeBuilds\4f482441-8b72-4531-987a-ff95273c27bd\install\x64-Debug (default)"  -DCMAKE_CXX_COMPILER="C:/Program Files (x86)/Microsoft Visual Studio/2017/Community/VC/Tools/MSVC/14.16.27023/bin/HostX64/x64/cl.exe"  -DCMAKE_C_COMPILER="C:/Program Files (x86)/Microsoft Visual Studio/2017/Community/VC/Tools/MSVC/14.16.27023/bin/HostX64/x64/cl.exe"  -DCMAKE_BUILD_TYPE="Debug" -DCMAKE_MAKE_PROGRAM="C:\PROGRAM FILES (X86)\MICROSOFT VISUAL STUDIO\2017\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\Ninja\ninja.exe" "C:\Users\MySelf\Source\Repos\merbanan\rtl_433"
1> Working directory: C:\Users\MySelf\CMakeBuilds\4f482441-8b72-4531-987a-ff95273c27bd\build\x64-Debug (default)
1> -- Found Git version: refs/heads/master commit 676a40426b0b9931cdf9991f60926182e8a06118 from 2023-01-18T21:58:32+0100
1> -- Using Git version tag: 22.11-40-g676a4042 on master at 202301182158
1> -- Could NOT find PkgConfig (missing: PKG_CONFIG_EXECUTABLE) 
1> -- IPv6 support enabled.
1> -- Threads support will be compiled.
1> -- Could NOT find OpenSSL, try to set the path to OpenSSL root folder in the system variable OPENSSL_ROOT_DIR (missing: OPENSSL_CRYPTO_LIBRARY OPENSSL_INCLUDE_DIR) 
1> -- OpenSSL development files not found, TLS won't be possible.
1> -- Could NOT find PkgConfig (missing: PKG_CONFIG_EXECUTABLE) 
1> -- Could NOT find PkgConfig (missing: PKG_CONFIG_EXECUTABLE) 
1> -- RTL-SDR device input will be compiled. Found version 
1> -- libusb-1.0 error messages are available. Found version 
1> -- SoapySDR device input will be compiled. Found version 0.8.1-PothosSDR-2021.07.25-vc16-x64
1> -- Could NOT find Doxygen (missing: DOXYGEN_EXECUTABLE) 
1> -- Configuring done
1> -- Generating done
1> -- Build files have been written to: C:/Users/MySelf/CMakeBuilds/4f482441-8b72-4531-987a-ff95273c27bd/build/x64-Debug (default)
1> Starting CMake target info extraction ...
1> CMake server connection made.
1> Extracted includes paths.
1> Extracted CMake variables.
1> Extracted source files and headers.
1> Extracted global settings.
1> Extracted code model.
1> Extracted CTest info.
1> Collating data ...
1> Target info extraction done.
No error were seen. 

Tried to run rtl_433.exe (C:\Users\MySelf\CMakeBuilds\e042b05f-1c24-8137-8705-1a9a735de5b7\build\x64-Debug (default)\src\) from VS 2017 “Select Startup Item” tab which opened a console window and multiple error messages( see attached images)
 
 Help, please
 
 
 
![Error 1 1-19-23](https://user-images.githubusercontent.com/3932952/213487419-b3995909-07f4-4474-8523-051d8627d53c.jpg)
![Error 2 1-19-23](https://user-images.githubusercontent.com/3932952/213487447-3b4da5fc-79fc-4262-ae15-8a6f78b82f2b.jpg)
![Error 3 1-19-23](https://user-images.githubusercontent.com/3932952/213487487-2f699985-e3ab-4d07-941d-3d8f9a2c81d3.jpg)
![Error 4 1-19-23](https://user-images.githubusercontent.com/3932952/213487507-d895c7db-0b5b-49f8-a9fe-9e146724aa4e.jpg)

**Comment by zuckschwerdt** [comment]
Did you maybe mix up files from the MinGW-w64 build with the Pothos install? Seems like headers to build against rtlsdr/Soapy are found but not the DLLs?

**Comment by MaxFun** [comment]
Don' think I did that. The files I downloaded for MinGW-w64 are in a different folder on the C drive and the Pothos install exe was downloaded to my download folded on C drive and run from there.

**Reply by zuckschwerdt** [reply]
Ok, there is a "search path" for DLLs and the Pothos installer amends to that (unless deselected maybe), is that path perhaps not correctly set up? You can try to copy the missing DLLs directly into that binaries folder, if that works then the search path is the problem.

**Comment by MaxFun** [comment]
Where do I find this ""search path" for DLLs" ?

**Reply by zuckschwerdt** [reply]
System Control Panel -> Advanced system settings -> Environment Variables -> `PATH` should be the one?

**Reply by MaxFun** [reply]
OK, in my  Environment Variables path I found the following for Pothos (see attachment)
Should I be looking something other than Pothos?

![Environment Variables path 1-19-23](https://user-images.githubusercontent.com/3932952/213512663-36f8858b-4532-4972-90c6-50d36df1faa0.jpg)

**Reply by zuckschwerdt** [reply]
The environment variable that is consulted form library (binaries) search is named `PATH` -- you'd want the Pothos dir(s) containing the DLLs to be listed there (`PATH`).

**Reply by MaxFun** [reply]
Sorry, but I do not understand. To find or add a system environment variable, I would need to know the variable name. Found the DLL  soapysdr.dll and rtlsdr.dll in C:\Program Files\PothosSDR\bin, so what would their variable name(s) be? The DLL msys-2.0.dll was not found in Potos directory

**Reply by MaxFun** [reply]
Ok, I think I understand, I added the paths to the DLL onto the existing  environment variable named Path. Still get the same error messages but will restart my computer to ensure the paths were added.

**Reply by zuckschwerdt** [reply]
Oh, `msys-2.0.dll` -- that's a smoking gun. It's a MinGW-w64 runtime library, not related to VS / MSVC… You are running the wrong binary, Mingw-w64 vs. VS. And you wrote that too:

> Build files have been written to: C:/Users/MySelf/CMakeBuilds/4f482441-8b72-4531-987a-ff95273c27bd/build/x64-Debug (default)

but

> Tried to run rtl_433.exe (C:\Users\MySelf\CMakeBuilds\e042b05f-1c24-8137-8705-1a9a735de5b7\build\x64-Debug (default)\src from VS 2017

Note that it's not the same path!

**Reply by MaxFun** [reply]
After restarting my computer, opening VS 2017, run rtl_433.exe (C:\Users\MySelf\CMakeBuilds\e042b05f-1c24-8137-8705-1a9a735de5b7\build\x64-Debug (default)\src) from VS 2017 “Select Startup Item” tab which opened a console window and a error message (see attachment)

It appears things got worse, VS out has multiple problems

'rtl_433.exe' (Win32): Loaded 'C:\Users\MySelf\CMakeBuilds\4f482441-8b72-4531-987a-ff95273c27bd\build\x64-Debug (default)\src\rtl_433.exe'. Symbols loaded.
'rtl_433.exe' (Win32): Loaded 'C:\Windows\System32\ntdll.dll'. Cannot find or open the PDB file.
'rtl_433.exe' (Win32): Loaded 'C:\Windows\System32\kernel32.dll'. Cannot find or open the PDB file.
'rtl_433.exe' (Win32): Loaded 'C:\Windows\System32\KernelBase.dll'. Cannot find or open the PDB file.
'rtl_433.exe' (Win32): Loaded 'C:\Windows\System32\ws2_32.dll'. Cannot find or open the PDB file.
'rtl_433.exe' (Win32): Loaded 'C:\Program Files\mingw-w64\x86_64-8.1.0-posix-seh-rt_v6-rev0\mingw64\bin\libusb-1.0.dll'. Module was built without symbols.
'rtl_433.exe' (Win32): Loaded 'C:\Windows\System32\rpcrt4.dll'. Cannot find or open the PDB file.
'rtl_433.exe' (Win32): Loaded 'C:\Windows\System32\shell32.dll'. Cannot find or open the PDB file.
'rtl_433.exe' (Win32): Loaded 'C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\IDE\CommonExtensions\Microsoft\TeamFoundation\Team Explorer\Git\usr\bin\msys-2.0.dll'. Module was built without symbols.
'rtl_433.exe' (Win32): Unloaded 'C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\IDE\CommonExtensions\Microsoft\TeamFoundation\Team Explorer\Git\usr\bin\msys-2.0.dll'
'rtl_433.exe' (Win32): Loaded 'C:\Windows\System32\msvcp_win.dll'. Cannot find or open the PDB file.
'rtl_433.exe' (Win32): Loaded 'C:\Windows\System32\ucrtbase.dll'. Cannot find or open the PDB file.
'rtl_433.exe' (Win32): Loaded 'C:\Program Files\PothosSDR\bin\rtlsdr.dll'. Cannot find or open the PDB file.
'rtl_433.exe' (Win32): Loaded 'C:\Windows\System32\user32.dll'. Cannot find or open the PDB file.
'rtl_433.exe' (Win32): Loaded 'C:\Windows\System32\win32u.dll'. Cannot find or open the PDB file.
'rtl_433.exe' (Win32): Loaded 'C:\Program Files\PothosSDR\bin\SoapySDR.dll'. Cannot find or open the PDB file.
The thread 0x36b8 has exited with code 0 (0x0).
The thread 0x1294 has exited with code 0 (0x0).
The thread 0x8f0 has exited with code 0 (0x0).
The program '[5820] rtl_433.exe' has exited with code -1073741701 (0xc000007b).

![Error 6  1-19-23](https://user-images.githubusercontent.com/3932952/213528528-c8b4345e-a428-4710-b644-76ebe01bc516.jpg)

**Reply by MaxFun** [reply]
There is no "C:/Users/MySelf/CMakeBuilds/4f482441-8b72-4531-987a-ff95273c27bd" folder on my computer  ?

**Reply by zuckschwerdt** [reply]
You are running a debug build and those (system) DLLs don't have a PDB, the messages about that can be ignored.

Sorry, that's likely all I can offer. I don't use Windows or VS even.

**Reply by MaxFun** [reply]
No sure what happened, searched again for C:\Users\MySelf\CMakeBuilds\4f482441-8b72-4531-987a-ff95273c27bd and found it.

**Comment by MaxFun** [comment]
Thanks for the help.

**Comment by MaxFun** [comment]
Got the built rtl_433.exe to run by doing the following
Add C:\Program Files\PothosSDR\bin to the System Environmental Variable path, edited Path and added. Even though I choose "Add PothosSDR to the system PATH for the current user" when installing PothosSDR, it did not happen.

Found rtl_433.exe with a timestamp with the time I ran VS 2017 and add to the folder it was in the following DLLs from the "C:\Program Files\PothosSDR\bin”

         rtlsdr.dll
    	libusb-1.0.dll
    	SoapySDR.dll

Opened the mingw-w64.terminal, cd to the located of the rtl_433.exe (C:\Users\MySelf\CMakeBuilds\e042b05f-1c24-8137-8705-1a9a735de5b7\install\x64-Debug (default)\bin>)
Entered “rtl_433.exe” and ran. I began receiving sensor data. 
It would be nice to have VS 2017 put the rtl_433.exe file into a bin folder in the VS’s build folder.
Such as “C:\Users\MySelf\source\repos\merbanan\rtl_433\bin”
I would think that modification of the CMakeLists.txt could do this.
![Output of code 1-22-23](https://user-images.githubusercontent.com/3932952/213917009-bd1ea0e3-c248-4e59-b58c-f5eb8f90f958.jpg)
![System Environment Path Edit 1-22-23](https://user-images.githubusercontent.com/3932952/213917021-ad7322b1-4337-412f-9099-1de001f73ecb.jpg)

**Comment by MaxFun** [comment]
Correction; I looked in the wrong Path, I looked in System Environment Variable path and PothosSDR was not there. It is in my user's path

**Reply by zuckschwerdt** [reply]
Thanks for the update, this will help others.
